force_bdss/data_sources/data_source_utilities.py

```diff
@@ -0,0 +1,196 @@
+        if not trait_similarity_check(
+                source, target,
+                attr_name, ignore_default):
+            failed_attr.append(attr_name)
+
+    return failed_attr
+
+
+def merge_trait_with_check(source, target, name, attributes=None,
+                           ignore_default=False):
+    """Performs the same function as `merge_trait` method, but with
+    some extra logic checks on default values and any selected
+    attributes supplied by `attributes` argument. These checks are
+    designed to determine whether both objects possess a similar
+    enough state to merge the `name` attribute.
+
+    Parameters
+    ----------
+    source, target: HasTraits
+        HasTraits instances to have an attribute synchronised
+    name: str
+        Name of an attribute on both source and target to be synced
+    attributes: str or list of str, optional
+        An attribute or list of attributes to check equivalence on both
+        source and target objects
+    ignore_default: bool, optional, default: False
+        Whether or not to ignore any attributes on source or target
+        that are are equal to their default values
+
+    Raises
+    -----
+    TraitSimilarityError, if any attribute checks fail
+    """
+
+    # Obtain names of any attributes in `attr_check` on both source and
+    # target that fail a `trait_similarity_check`
+    failed_attr = attr_checker(source, target, attributes,
+                               ignore_default=ignore_default)
+
+    if failed_attr:
+        attr_name = failed_attr[0]
+        error_msg = (
+            "Source object has failed a trait "
+            "similarity check with target: "
+            "The {} attribute of source ({}) doesn't match "
+            "target ({}).".format(
+                attr_name,
+                getattr(source, attr_name),
+                getattr(target, attr_name)
+            )
+        )
+        logger.exception(error_msg)
+        raise TraitSimilarityError(error_msg)
+
+    # Merge attribute `name` between source and target
+    merge_trait(source, target, name)
+
+
+def merge_lists_with_check(new_list, old_list, attributes):
+    """Perform a merge_trait_with_check of each attribute in
+    `attributes` between the corresponding elements of new_list
+    and old_list. Overwrite priority is given to the values in
+    new_list and all current attributes must be similar enough
+    to pass a trait_similarity_check
+
+    Parameters
+    ----------
+    new_list: list
+        List of new elements
+    old_list: list
+        List of old elements to be retained
+    attributes: str or list of str, optional
+        An attribute or list of attributes to both merge and
+        check similarity on each corresponding element in
+        new_list and old_list
+    ignore_default: bool, optional, default: False
+        Whether or not to ignore any attributes on source or target
+        that are are equal to their default values
```

Comment:
`merge_lists_with_check` doesn't have the `ignore_default` parameter

Reply:
Good spot